ADR-042: Group module (#9089)
cosmos/cosmos-sdk: docs/architecture/adr-042-group-module.md

    uint64 group_id = 1;

    // admin is the account address of the group's admin.
    string admin = 2;
    
    // metadata is any arbitrary metadata to attached to the group.
    bytes metadata = 3;

    // version is used to track changes to a group's membership structure that
    // would break existing proposals. Whenever a member weight has changed,
    // or any member is added or removed, the version is incremented and will
    // invalidate all proposals from older versions.
    uint64 version = 4;

    // total_weight is the sum of the group members' weights.
    string total_weight = 5;
}
```

```proto
message GroupMember {

    // group_id is the unique ID of the group.
    uint64 group_id = 1;

    // member is the member data.
    Member member = 2;
}

// Member represents a group member with an account address,
// non-zero weight and metadata.
message Member {

    // address is the member's account address.
    string address = 1;
    
    // weight is the member's voting weight that should be greater than 0.
    string weight = 2;
    
    // metadata is any arbitrary metadata to attached to the member.
    bytes metadata = 3;
}
```

> ### Group Account

Author: In the current implementation, only the current group admin is authorized to create a group account for the group.

    google.protobuf.Any decision_policy = 6 [(cosmos_proto.accepts_interface) = "DecisionPolicy"];
}
```

Similarly to a group admin, a group account admin can update its metadata, decision policy or set a new group account admin.

A group account can also be an admin or a member of a group.
For instance, a group admin could be another group account which could "elects" the members or it could be the same group that elects itself.

### Decision Policy

A decision policy is the mechanism by which members of a group can vote on 
proposals.

All decision policies should have a minimum and maximum voting window.
The minimum voting window is the minimum duration that must pass in order
for a proposal to potentially pass, and it may be set to 0. The maximum voting
window is the maximum time that a proposal may be voted on and executed if 
it reached enough support before it is closed.
Both of these values must be less than a chain-wide max voting window parameter.

We define the `DecisionPolicy` interface that all decision policies must implement:

```go
type DecisionPolicy interface {
	codec.ProtoMarshaler

	ValidateBasic() error
	GetTimeout() types.Duration

> 	Allow(tally Tally, totalPower string, votingDuration time.Duration) (DecisionPolicyResult, error)

Author: afaik, we've only had the `ThresholdDecisionPolicy`.
I guess that for quadratic voting, it wouldn't be only about the decision policy but also how the members weights are managed.

> 	Allow(tally Tally, totalPower string, votingDuration time.Duration) (DecisionPolicyResult, error)

Reviewer: Is it possible for a `DecisionPolicy` to keep some state or is it ideal for it to be stateless. I guess another way of putting this is did you guys consider instead of giving the decision policy the entire tally that you just give it the vote. In this regard the decision policy would manage the tally

# ADR 042: Group Module

## Changelog

- 2020/04/09: Initial Draft

## Status

Draft

## Abstract

This ADR defines the `x/group` module which allows the creation and management of on-chain multi-signature accounts and enables voting for message execution based on configurable decision policies.

## Context

The current multi-signature mechanism of the Cosmos SDK has certain limitations:
- Key rotation is not possible, although this can be solved with [account rekeying](adr-034-account-rekeying.md).
- Thresholds can't be changed.

> - UX is not straightforward for non-technical users ([#5661](https://github.com/cosmos/cosmos-sdk/issues/5661)).

Author: ```suggestion
- UX is cumbersome for non-technical users ([#5661](https://github.com/cosmos/cosmos-sdk/issues/5661)).
```

- It requires `legacy_amino` sign mode ([#8141](https://github.com/cosmos/cosmos-sdk/issues/8141)).

While the group module is not meant to be a total replacement for the current multi-signature accounts, it provides a solution to the limitations described above, with a more flexible key management system where keys can be added, updated or removed, as well as configurable thresholds.

> It's meant to be used with other key management modules such as [`x/feegrant`](./adr-029-fee-grant-module.md) ans [`x/authz`](adr-030-authz-module.md) to simplify key management for individuals and organizations.

Author: I was just using the same terminology as this existing meta-issue https://github.com/cosmos/cosmos-sdk/issues/7074

> It's meant to be used with other key management modules such as [`x/feegrant`](./adr-029-fee-grant-module.md) ans [`x/authz`](adr-030-authz-module.md) to simplify key management for individuals and organizations.

Reviewer 2: It seams like the meta issue is more general and want to talks about different authorization schemes, rather than just key management.

> It's meant to be used with other key management modules such as [`x/feegrant`](./adr-029-fee-grant-module.md) ans [`x/authz`](adr-030-authz-module.md) to simplify key management for individuals and organizations.

Author: ```suggestion
It's meant to be used with other access control modules such as [`x/feegrant`](./adr-029-fee-grant-module.md) ans [`x/authz`](adr-030-authz-module.md) to simplify key management for individuals and organizations.
```

> The current implementation of the group module can be found in https://github.com/regen-network/regen-ledger/tree/master/proto/regen/group/v1alpha1 and https://github.com/regen-network/regen-ledger/tree/master/x/group.

Reviewer 3: This code implements adr033, correct? 

Another note, the coding style could cause some confusion since we have a recommended module layout, the groups module implements something new that seems to lack documentation as well.

Author: yes indeed, we're currently having a broader conversion about module wiring here: https://github.com/regen-network/regen-ledger/issues/301 so this needs to be figured out before implementing the group module as part of the sdk

> The current implementation of the group module can be found in https://github.com/regen-network/regen-ledger/tree/master/proto/regen/group/v1alpha1 and https://github.com/regen-network/regen-ledger/tree/master/x/group.

Author: yeah strictly speaking this is not the first version so maybe let's use proof of concept

> The current implementation of the group module can be found in https://github.com/regen-network/regen-ledger/tree/master/proto/regen/group/v1alpha1 and https://github.com/regen-network/regen-ledger/tree/master/x/group.

Author: ```suggestion
The proof of concept of the group module can be found in https://github.com/regen-network/regen-ledger/tree/master/proto/regen/group/v1alpha1 and https://github.com/regen-network/regen-ledger/tree/master/x/group.
```

## Decision

> We propose merging the `x/group` module with its supporting [ORM/Table Store package](https://github.com/regen-network/regen-ledger/tree/master/orm) ([#7098](https://github.com/cosmos/cosmos-sdk/issues/7098)) into the Cosmos SDK and continuing development here. There will be a dedicated ADR for the ORM package.

Reviewer 3: will the ADR come before or after its merged?

Author: Ideally it should come before, @i-norden and I will likely coordinate on this

### Group

A group is an aggregation of accounts with associated weights. It is not
an account and doesn't have a balance. It doesn't in and of itself have any

> sort of voting or decision weight. It has an `admin` account which can manage members in the group.

Reviewer 2: Oh, I have an answer - this is explained in the `GroupAccount` section. For having an easier read, could you make a paragraph before the `Group` section to roughly talk about Group and GroupAccount (what's the use-case for each and what's the relationship between them)?

> sort of voting or decision weight. It has an `admin` account which can manage members in the group.

Author: > > It doesn't in and of itself have any sort of voting or decision weight.
> 
> Not sure what do you mean with that? Group account can be used in another group account (and vote), right?
> 
> Why we don't allow keepeing a balance for a group account? It has an address and it can use `x/bank` , right?

We're talking about groups here, not group accounts.

> sort of voting or decision weight. It has an `admin` account which can manage members in the group.

Author: > Oh, I have an answer - this is explained in the `GroupAccount` section. For having an easier read, could you make a paragraph before the `Group` section to roughly talk about Group and GroupAccount (what's the use-case for each and what's the relationship between them)?

ok I'll try to make this clearer

> sort of voting or decision weight. It has an `admin` account which can manage members in the group.

Author: yes this can be done by having a group account as the  group `admin`.

> sort of voting or decision weight. It has an `admin` account which can manage members in the group.

Reviewer 2: Let's add a note about this as well.

> Groups are stored in state as part of an ORM-based `groupTable` using the `GroupInfo` type. The `group_id` is an auto-increment integer.

Reviewer 3: It seems this module is reliant on the ORM module. It would be good to first understand and merge that code then this code IMO.

Author: agreed, see https://github.com/cosmos/cosmos-sdk/pull/9089#discussion_r613071227

```proto
message GroupInfo {

>     // group_id is the unique ID of the group.

Reviewer 2: ```suggestion
    // group_id is the unique ID of this group.
```

    uint64 group_id = 1;

    // admin is the account address of the group's admin.
    string admin = 2;
    
    // metadata is any arbitrary metadata to attached to the group.
    bytes metadata = 3;

    // version is used to track changes to a group's membership structure that
    // would break existing proposals. Whenever any members weight is changed,
    // or any member is added or removed this version is incremented and will
    // cause proposals based on older versions of this group to fail
    uint64 version = 4;

    // total_weight is the sum of the group members' weights.
    string total_weight = 5;
}
```

> Group members are stored in a `groupMemberTable` using the `GroupMember` type:

Author: it's a `Table` from the orm package yes.
I'll move orm related stuff to a specific paragraph https://github.com/cosmos/cosmos-sdk/pull/9089#discussion_r613106500

```proto
message GroupMember {

    // group_id is the unique ID of the group.
    uint64 group_id = 1;

    // member is the member data.
    Member member = 2;
}

// Member represents a group member with an account address,
// non-zero weight and metadata.
message Member {

    // address is the member's account address.

>     string address = 1;

Author: Yes it's theoretically possible, same for the group admin account.

>     string address = 1;

Reviewer 4: Or module addresses?

>     string address = 1;

Author: A group account is a module account itself so this works in this case but how could a regular module account vote?

>     string address = 1;

Reviewer 4: Might want to create an automated voting module or offload to cosmwasm.

    
    // weight is the member's voting weight that should be greater than 0.
    string weight = 2;
    
    // metadata is any arbitrary metadata to attached to the member.

>     bytes metadata = 3;

Author: Here's a use case in the context of regen-ledger: this could represent a hash of some data from the [`x/data` module](https://docs.regen.network/modules/data/) which is used for timestamping, signing, and storing data.

}
```

### Group Account

A group account is an account associated with a group and a decision policy.
A group account does have a balance.

Group accounts are abstracted from groups because a single group may have
multiple decision policies for different types of actions. Managing group
membership separately from decision policies results in the least overhead
and keeps membership consistent across different policies. The pattern that

> is recommended is to have a single master group account for a given group,

Reviewer 5: This pattern is very nice to highlight

and then to create separate group accounts with different decision policies
and delegate the desired permissions from the master account to
those "sub-accounts" using the [`x/authz` module](adr-030-authz-module.md).

> Group accounts are stored as part of the `groupAccountTable` and modeled by `GroupAccountInfo`.

Reviewer 2: ```suggestion
Group accounts are stored as part of the `groupAccountTable` and defined by `GroupAccountInfo`.
```

> Group accounts are stored as part of the `groupAccountTable` and modeled by `GroupAccountInfo`.

Reviewer 2: `groupAccountTable` is specific to `orm`. So, let's move that details to the ORM section.

```proto

> message GroupAccountInfo {

Reviewer 2: Why do we have `Info` suffix? Based on the paragraph above, I would call this `PolicyAccount`.

Author: It was in order to be consistent with x/ecocredit from regen-ledger which uses this suffix too, but I'd be ok with dropping it.
We use `Group` prefix because it's related to a given group and I believe it feels more natural as part of the `x/group` module.

Reviewer 2: So, how about `PolicyGroup`?

Author: But then you lose info about the fact that it's an actual account. I kinda like the current `GroupAccount` naming but don't have a strong preference either. It'd be good to have some thoughts from @aaronc who's the primarily designer of the group module afaik.

Reviewer 2: Then, how about `PolicyAccount`? I don't think we need to put `Group` everywhere.

    // address is the group account address.
    string address = 1;
    

>     // group_id is the unique ID of the group.

Reviewer 2: ```suggestion
    // group_id is the ID of the Group the PolicyAccount  belongs to.
```

    uint64 group_id = 2;

    // admin is the account address of the group admin.

>     string admin = 3;

Reviewer 4: I thought this could be achieved by setting the admin as a group account with all members.

>     string admin = 3;

Author: the group admin can also update the group, for instance updating metadata or setting a new admin (same for the group account admin, ie can update metadata, decision policy or set a new group account admin)

>     string admin = 3;

Reviewer 6: The admin could another group account which "elects" the members or it could be the same group so that it elects itself.

>     // metadata is any arbitrary metadata to attached to the group account.

Reviewer 2: ```suggestion
    // metadata is any arbitrary metadata of this group account.
```

    bytes metadata = 4;

>     // version is used to track changes to a group's GroupAccountInfo structure that
>     // would create a different result on a running proposal.

Reviewer 2: ```suggestion
    // version is used to track changes to a group's GroupAccountInfo structure that
    // invalidates active proposal from old versions.
```

    uint64 version = 5;

    // decision_policy specifies the group account's decision policy.
    google.protobuf.Any decision_policy = 6 [(cosmos_proto.accepts_interface) = "DecisionPolicy"];
}
```

> The group account address is generated based on an auto-increment integer which is used to derive the group module `RootModuleKey` into a `DerivedModuleKey`, as stated in [ADR-033](adr-033-protobuf-inter-module-comm.md#modulekeys-and-moduleids). The group account is added as a new `ModuleAccount` through `x/auth`.

Reviewer 2: ditto, let's move it to an ORM section. It's an implementation detail.

### Decision Policy

> A decision policy is the mechanism by which members of a group can vote on 
> proposals.

Reviewer 2: ```suggestion
A decision policy is an interface for voting policies for group members.
```

Author: I find this suggestion is a bit redundant "A decision **policy** is an interface for voting **policies**..."

> All decision policies generally would have a minimum and maximum voting window.

Reviewer 2: ```suggestion
All decision policies should have a minimum and maximum voting window.
```

> The minimum voting window is the minimum amount of time that must pass in order
> for a proposal to potentially pass, and it may be set to 0. The maximum voting
> window is the maximum time that a proposal may be voted on before it is closed.
> Both of these values must be less than a chain-wide max voting window parameter.

Reviewer 2: ```suggestion
Voting for each proposal is possible for at least the minimum voting window (can be 0) 
and not more than the maximum voting window. Proposal can be executed after the 
minimum voting window if it reached enough support. 
Both of these values must be less than a chain-wide max voting window parameter.
```

Author: I find the suggestion a bit confusing and prefer the original phrasing with 2 different sentences: one to explain the min voting window and another one for the max voting window. But I can add something about proposal execution:
```suggestion
The minimum voting window is the minimum amount of time that must pass in order
for a proposal to potentially pass, and it may be set to 0. The maximum voting
window is the maximum time that a proposal may be voted on and executed if 
it reached enough support before it is closed.
Both of these values must be less than a chain-wide max voting window parameter.
```

Reviewer 2: Looks better. I would change _minimum amount of time_ to _minimum duration_

I don't like this kind of sentences (from original version): _The minimum voting window is the minimum amount of time that must pass in order for a proposal to potentially pass, and it may be set to 0._  - it's too much repetition and too complex to read.

Author: ok updated

We define the `DecisionPolicy` interface that all decision policies must implement:

```go
type DecisionPolicy interface {
	codec.ProtoMarshaler

	ValidateBasic() error
	GetTimeout() types.Duration

> 	Allow(tally Tally, totalPower string, votingDuration time.Duration) (DecisionPolicyResult, error)

Reviewer: Ah I just read further down. Is it calculated whenever someone sends `Msg/Exec`?

> 	Allow(tally Tally, totalPower string, votingDuration time.Duration) (DecisionPolicyResult, error)

Author: > Will a proposal be executed the moment that allow is true?

A separate tx with `Msg/Exec` has to be explicitely sent for the proposal to be executed for now.

>Ah I just read further down. Is it calculated whenever someone sends Msg/Exec?

It's calculated on new votes and on `Msg/Exec`

	Validate(g GroupInfo) error
}

type DecisionPolicyResult struct {
	Allow bool

> 	Final bool

Author: When set to `true`, it means that `Allow` represents the final result of the tally.

}
```

#### Threshold decision policy

> A threshold decision policy defines a threshold of yes votes (based on a tally
> of voter weights) that must be achieved in order for a proposal to pass. For
> this decision policy, abstain and veto are simply treated as no's.

Reviewer 2: ```suggestion
A threshold decision policy defines a minimum support votes (_yes_), based on a tally
of voter weights, for a proposal to pass. Abstain and veto are treated as no support (_no_).
```

```proto
message ThresholdDecisionPolicy {

>     // threshold is the minimum weighted sum of yes votes that must be met or exceeded for a proposal to succeed.

Reviewer 2: ```suggestion
    // threshold is the minimum weighted sum of support votes for a proposal to succeed.
```

    string threshold = 1;
    
    // timeout is the duration from submission of a proposal to the end of voting period
    // Within this times votes and exec messages can be submitted.

>     google.protobuf.Duration timeout = 2 [(gogoproto.nullable) = false];

Reviewer 2: timeout sounds more like a specific time, not a duration. How about calling it `voting_window`?

Reviewer: I agree that `voting_window` or `voting_period` would be more intuitive

}
```

### Proposal

> Any member of a group can submit a proposal for a group account to decide upon.

Reviewer 2: ```suggestion
Any member of a group can submit a proposal to a group account to vote and execute.
```

Author: This is not entirely true, anyone can execute a proposal.

A proposal consists of a set of `sdk.Msg`s that will be executed if the proposal
passes as well as any metadata associated with the proposal.

> Proposals are stored as part of the `proposalTable` using the `Proposal` type. The `proposal_id` is an auto-increment integer.

Reviewer 2: ditto

        
        // An empty value is invalid and not allowed.
        STATUS_UNSPECIFIED = 0 [(gogoproto.enumvalue_customname) = "ProposalStatusInvalid"];
        
        // Initial status of a proposal when persisted.
        STATUS_SUBMITTED = 1 [(gogoproto.enumvalue_customname) = "ProposalStatusSubmitted"];
        
        // Final status of a proposal when the final tally was executed.
        STATUS_CLOSED = 2 [(gogoproto.enumvalue_customname) = "ProposalStatusClosed"];
        
        // Final status of a proposal when the group was modified before the final tally.
        STATUS_ABORTED = 3 [(gogoproto.enumvalue_customname) = "ProposalStatusAborted"];
    }

    // Status represents the high level position in the life cycle of the proposal. Initial value is Submitted.
    Status status = 8;

    // Result defines types of proposal results.
    enum Result {
        option (gogoproto.goproto_enum_prefix) = false;

        // An empty value is invalid and not allowed
        RESULT_UNSPECIFIED = 0 [(gogoproto.enumvalue_customname) = "ProposalResultInvalid"];
        
        // Until a final tally has happened the status is unfinalized

>         RESULT_UNFINALIZED = 1 [(gogoproto.enumvalue_customname) = "ProposalResultUnfinalized"];

Reviewer 2: ```suggestion
        RESULT_NOTFINALIZED = 1 [(gogoproto.enumvalue_customname) = "ProposalResultNotfinalized"];
```

Author: not sure why this is better? Also this would need to be `RESULT_NOT_FINALIZED` in 2 words instead which is longer and does not bring more info

       
        // Final result of the tally
        RESULT_ACCEPTED = 2 [(gogoproto.enumvalue_customname) = "ProposalResultAccepted"];
        
        // Final result of the tally
        RESULT_REJECTED = 3 [(gogoproto.enumvalue_customname) = "ProposalResultRejected"];
    }

    // result is the final result based on the votes and election rule. Initial value is unfinalized.
    // The result is persisted so that clients can always rely on this state and not have to replicate the logic.
    Result result = 9;

    // vote_state contains the sums of all weighted votes for this proposal.
    Tally vote_state = 10 [(gogoproto.nullable) = false];

>     // timeout is the timestamp of the block where the proposal execution times out. Header times of the votes and execution messages
>     // must be before this end time to be included in the election. After the timeout timestamp the proposal can not be
>     // executed anymore and should be considered pending delete.

Reviewer 2: ```suggestion
    // timeout is the block timestamp deadline for proposal. All votes and execution messages must be before processed before the timeout. After the timeout, the proposal can not be executed anymore and should be considered pending delete.
```

>     google.protobuf.Timestamp timeout = 11 [(gogoproto.nullable) = false];

Reviewer 2: so, here we use `timeout` correctly - it's not a duration.

    // ExecutorResult defines types of proposal executor results.
    enum ExecutorResult {
        option (gogoproto.goproto_enum_prefix) = false;
        
        // An empty value is not allowed.
        EXECUTOR_RESULT_UNSPECIFIED = 0  [(gogoproto.enumvalue_customname) = "ProposalExecutorResultInvalid"];
        
        // We have not yet run the executor.
        EXECUTOR_RESULT_NOT_RUN = 1 [(gogoproto.enumvalue_customname) = "ProposalExecutorResultNotRun"];
        
        // The executor was successful and proposed action updated state.
        EXECUTOR_RESULT_SUCCESS = 2 [(gogoproto.enumvalue_customname) = "ProposalExecutorResultSuccess"];
        
        // The executor returned an error and proposed action didn't update state.
        EXECUTOR_RESULT_FAILURE = 3 [(gogoproto.enumvalue_customname) = "ProposalExecutorResultFailure"];
    }

    // executor_result is the final result based on the votes and election rule. Initial value is NotRun.
    ExecutorResult executor_result = 12;

    // msgs is a list of Msgs that will be executed if the proposal passes.

>     repeated google.protobuf.Any msgs = 13;

Reviewer: Do these get validated somehow before the proposal is actually made?

Author: Yes, proposal `Msgs` get validated as part of the `MsgCreateProposalRequest` validation.

Author: added to Proposal section

}

// Tally represents the sum of weighted votes.
message Tally {
    option (gogoproto.goproto_getters) = false;

    // yes_count is the weighted sum of yes votes.
    string yes_count = 1;
    
    // no_count is the weighted sum of no votes.
    string no_count = 2;
    
    // abstain_count is the weighted sum of abstainers.
    string abstain_count = 3;
    
    // veto_count is the weighted sum of vetoes.

>     string veto_count = 4;

Reviewer 2: for the definition above, the `veto` doesn't have any extra functionality, and it's equal to `no`. So why do we have it?

Author: it could be useful for certain decision policies.

}

> ```

Reviewer: I think it would really open up governance possibilities if we supported non-binary choices or at least weren't fixed with the choices that someone could make. I'm thinking of something along the lines of:
```proto
message Tally {
    // schema represents how the choices should be interpreted 
    // (maybe this goes in the proposal or is somehow known by the decision policy)
    int64 schema = 1; 

    // choices is ordered identically to the members in the group so as 
    // to easily match the member weighting with the choice
    repeated Choice choices = 2; 
}
```
Where a `Choice` is just:
```go
type Choice uint64
```

Reviewer 6: Yeah, we should think of ways to make it more flexible. My thinking was that yes, no, veto and abstain is a fairly common pattern for binary voting and a good starting point. But it would be cool if we supported other patterns like ranked choice, etc.

Author: added to the list of "Further Discussions"

    Choice choice = 3;

    // metadata is any arbitrary metadata to attached to the vote.
    bytes metadata = 4;

    // submitted_at is the timestamp when the vote was submitted.
    google.protobuf.Timestamp submitted_at = 5 [(gogoproto.nullable) = false];
}
```

Voting internally updates the proposal `Status` and `Result` if needed.

### Executing Proposals

Proposals will not be automatically executed by the chain in this current design,
but rather a user must submit a `Msg/Exec` transaction to attempt to execute the
proposal based on the current votes and decision policy. A future upgrade could
automate this and have the group account (or a fee granter) pay.

Inter-module communication introduced by [ADR-033](adr-033-protobuf-inter-module-comm.md) will be used to route a proposal's messages using the `DerivedModuleKey` corresponding to the proposal's group account. It can also temporarily support routing of non `ServiceMsg`s through the `sdk.Router` (see [#8864](https://github.com/cosmos/cosmos-sdk/issues/8864)).
For these messages to execute successfully, their signer should be set as the group account.

#### Changing Group Membership

> In the current implementation, updating a group or a group account after submitting a proposal will make it invalid. It will simply fail if someone calls `Msg/Exec` and will eventually be garbage collected.

Reviewer: I feel this adds a great deal of inflexibility. If for example a proposal lasts two weeks, then not only must arriving and leaving members wait until the proposal is finished but member weights can not be updates until after the proposal is complete. I'm guessing a default way of assigning weights is based on the amount of tokens a member contributes. By freezing this, I feel we take away a lot of sovereignty for group members and usability of groups. This will be even more problematic in large groups with potentially thousands of members.

Have you considered taking a snapshot of the member weights at the moment the proposal is submitted and using that as the reference for the decision policy? Perhaps this is more of a future improvement

Reviewer 5: I agree that this is very inflexible. It works quite well for the multisig use case, or even very-infrequently-updated committees. But is completely unsuited for frequently changing groups (like groups based on stake). 

However, it is much more efficient for this (quite useful) use case. And adding snapshotting to a company multisig might be overkill. It should be clear what the scope of the group module is.

Reviewer: Right. That makes sense in the case where groups hardly change. I guess I was coming more from the angle of governance. Perhaps this could be handled by the `DecisionPolicy`. That way groups pick a decision policy that matches their use case. For relatively static groups, the decision policy voids a proposal when the group changes. For more dynamic groups, the decision policy takes a snapshot of the weights of all members at the submission of the proposal

Reviewer 2: Group Account is a limited DAO contract. I don't think we should expect from the Group Account same complexity and same features as we would from a fully fledged DAO.

Reviewer 6: This whole issue is addressed more thoroughly in #9066. Briefly, I believe that when weights are calculated should be a pluggable part of `DecisionPolicy`. For many use cases, the final weights are ideal IMHO and tallying can simply be done at the end of voting. In other use cases, we need a snapshot at the start of voting and this brings other technical complexities we will need to address (see https://github.com/cosmos/cosmos-sdk/discussions/9066#discussioncomment-605745)

## Consequences

### Positive

- Improved UX for multi-signature accounts allowing key rotation and custom decision policies.

### Negative

### Neutral

- It uses ADR 033 so it will need to be implemented within the Cosmos SDK, but this doesn't imply necessarily any large refactoring of existing Cosmos SDK modules.

> - It requires the ORM package.

Reviewer 2: The group module doesn't require the ORM package. It's current implementation uses it. So let's rephrase it.

Author: ```suggestion
- The current implementation of the group module requires the ORM package.
```
wdyt about that @robert-zaremba?

Reviewer 2: Sounds good. I would use `uses` instead of `requires`.
